### PR TITLE
Check that file output directories exist

### PR DESCRIPTION
**Pull request by cpethrick**
Add logic to the parameters parsing sections to ensure that output directories exist. I chose to abort rather than create the missing directories to prevent cases where someone may have launched from the wrong working directory or otherwise made a mistake in their simulation setup. I could be persuaded to just create missing directories, though.

The change is motivated by the following `deal.ii` error message, which gives a correct but overly general diagnosis of the problem.
image

### CODE REVIEW

File: src/parameters/parameters_flow_solver.cpp [hunk 1]
```diff
@@ -408,6 +425,13 @@ void FlowSolverParam::parse_parameters(dealii::ParameterHandler &prm)
           number_of_times_to_output_velocity_field = get_number_of_values_in_string(output_velocity_field_times_string);
           output_vorticity_magnitude_field_in_addition_to_velocity = prm.get_bool("output_vorticity_magnitude_field_in_addition_to_velocity");
           output_flow_field_files_directory_name = prm.get("output_flow_field_files_directory_name");
+            // Check if directory exists - see https://stackoverflow.com/a/18101042
+            struct stat info;
+            if( stat( output_flow_field_files_directory_name.c_str(), &info ) != 0 ){
+                pcout << "Error: No flow field files directory named " << output_flow_field_files_directory_name << " exists." << std::endl
+                          << "Please create the directory and restart." << std::endl;
```
Comment (jbrillon) [inline src/parameters/parameters_flow_solver.cpp:432]:
```suggestion
                          << "Please create the directory and restart. Aborting..." << std::endl;
```

File: src/parameters/parameters_flow_solver.cpp [hunk 2]
```diff
@@ -334,6 +344,13 @@ void FlowSolverParam::parse_parameters(dealii::ParameterHandler &prm)
         restart_computation_from_file = prm.get_bool("restart_computation_from_file");
         output_restart_files = prm.get_bool("output_restart_files");
         restart_files_directory_name = prm.get("restart_files_directory_name");
+        // Check if directory exists - see https://stackoverflow.com/a/18101042
+        struct stat info;
+        if( stat( restart_files_directory_name.c_str(), &info ) != 0 ){
+            pcout << "Error: No restart directory named " << restart_files_directory_name << " exists." << std::endl
+                      << "Please create the directory and restart." << std::endl;
```
Comment (jbrillon) [inline src/parameters/parameters_flow_solver.cpp:351]:
```suggestion
                      << "Please create the directory and restart. Aborting..." << std::endl;
```

File: src/parameters/parameters_flow_solver.cpp [hunk 3]
```diff
@@ -334,6 +344,13 @@ void FlowSolverParam::parse_parameters(dealii::ParameterHandler &prm)
         restart_computation_from_file = prm.get_bool("restart_computation_from_file");
         output_restart_files = prm.get_bool("output_restart_files");
         restart_files_directory_name = prm.get("restart_files_directory_name");
+        // Check if directory exists - see https://stackoverflow.com/a/18101042
+        struct stat info;
+        if( stat( restart_files_directory_name.c_str(), &info ) != 0 ){
+            pcout << "Error: No restart directory named " << restart_files_directory_name << " exists." << std::endl
```
Comment (jbrillon) [inline src/parameters/parameters_flow_solver.cpp:350]:
```suggestion
            pcout << "Error: No restart files directory named " << restart_files_directory_name << " exists." << std::endl
```

File: src/parameters/all_parameters.cpp [hunk 4]
```diff
@@ -474,6 +478,13 @@ void AllParameters::parse_parameters (dealii::ParameterHandler &prm)
     if (non_physical_behavior_string == "print_warning")     { non_physical_behavior_type = NonPhysicalBehaviorEnum::print_warning;}
 
     solution_vtk_files_directory_name = prm.get("solution_vtk_files_directory_name");
+    // Check if directory exists - see https://stackoverflow.com/a/18101042
+    struct stat info;
+    if( stat( solution_vtk_files_directory_name.c_str(), &info ) != 0 ){
+        pcout << "Error: No solution vtk files directory named " << solution_vtk_files_directory_name << " exists." << std::endl
+                  << "Please create the directory and restart." << std::endl;
```
Comment (jbrillon) [inline src/parameters/all_parameters.cpp:485]:
```suggestion
                  << "Please create the directory and restart. Aborting..." << std::endl;
```

**Review by jbrillon — APPROVED** [review]
Great PR, very good check to have in the code. I've left a few code suggestions on the print statements

File: src/parameters/parameters_flow_solver.cpp [hunk 5]
```diff
@@ -408,6 +425,13 @@ void FlowSolverParam::parse_parameters(dealii::ParameterHandler &prm)
           number_of_times_to_output_velocity_field = get_number_of_values_in_string(output_velocity_field_times_string);
           output_vorticity_magnitude_field_in_addition_to_velocity = prm.get_bool("output_vorticity_magnitude_field_in_addition_to_velocity");
           output_flow_field_files_directory_name = prm.get("output_flow_field_files_directory_name");
+            // Check if directory exists - see https://stackoverflow.com/a/18101042
+            struct stat info;
```
Comment (dougshidong) [inline src/parameters/parameters_flow_solver.cpp:429]: Resolve `struct stat info;` shadowing local variable. 
https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=244&id=dougshidong_PHiLiP&open=AYr7U1ps4kO947chs77C

**Review by dougshidong — CHANGES_REQUESTED** [review]

File: src/parameters/parameters_flow_solver.cpp (hunk 5 above)
Comment (cpethrick) [reply src/parameters/parameters_flow_solver.cpp]: Fixed, thanks for highlighting that.

**Review by PranshulThakur — APPROVED** [review]
LGTM! Thanks for adding the checks.

**Review by dougshidong — APPROVED** [review]

**Pull request merged**